### DISCUSSION TITLE
get OPENAI_API_KEY from source other than os.getenv

### DISCUSSION THREAD

**Discussion by Dronakurl** [Robitx/gp.nvim#55] [Q&A] [answered]
I want to get the API key from a local file rather than an environment variable.
The following code is in my lazy.nvim configuration

```lua
local command = "cat " .. vim.fn.expand("$HOME") .. "/.config/openaikey"
local handle = io.popen(command)
API_KEY = nil
if handle ~= nil then
  API_KEY = handle:read("*a")
  handle:close()
end
return {
  {
    "robitx/gp.nvim",
    opts = { openai_api_key = API_KEY },
  }
}
```

.. gives this error message, whenever I try to submit a prompt. Just pasting the key (`openai_api_key = "jffjkkgkkk..."`) instead of the variable and also `print(API_KEY)` shows that the key was loaded correctly in the variable `API_KEY`.

```
/home/kk/.local/share/nvim/lazy/gp.nvim/lua/gp/init.lua:652: E5560: nvim_echo must not be called in a lua loop callback
stack traceback:
	[C]: in function 'nvim_echo'
	/home/kk/.local/share/nvim/lazy/gp.nvim/lua/gp/init.lua:652: in function 'error'
	/home/kk/.local/share/nvim/lazy/gp.nvim/lua/gp/init.lua:978: in function 'out_reader'
```

Could anyone please help me?

**Comment by Robitx** [comment, marked as answer]
@Dronakurl hey, try to strip the white spaces like a newline from the key
``` lua
	API_KEY = handle:read("*a"):match("%S+")
```

You might miss those using simple print =>
``` lua
print("API_KEY: " .. vim.inspect(API_KEY))
```

Also thanks, I forgot to wrap new error/warn/info messages (which use nvim_echo) into vim.schedule (that's why you saw E5560 instead of what actually happened). Fixed in the latest version.

**Comment by Dronakurl** [comment]
I got it to work, thank you very much for your help.
I am very new to lua, so I could not figure this out
